### ISSUE TITLE
Stream-Adventure : HTML & Trumpet  = Pipe confusion

### ISSUE THREAD

**Issue by joshmu** [nodeschool/discussions#346]
Feel as though I am missing something here... I couldn't possibly solve this problem as I don't understand the logic of piping back to trumpet?  Wouldn't that just create an endless loop?  I'm not alone as I found this [Stackoverflow question](http://stackoverflow.com/questions/24103981/how-does-piping-a-stream-back-to-itself-work-with-trumpet) on the exact same issue.

```
var trumpet = require('trumpet');
var through = require('through');
var tr = trumpet();

var loud = tr.select('.loud').createStream();
loud.pipe(through(function (buf) {
    this.queue(buf.toString().toUpperCase());
})).pipe(loud);

process.stdin.pipe(tr).pipe(process.stdout);
```

Could anyone enlighten me?

**Comment by JimVercoelen** [comment]
Hi all,

Just to be sure @davecocoa for understanding your explanation more:

'tr receives an input, and then sends output as a result.

loud sends an output, and waits for input to arrive as a result."

It works a bit like this?:
![yolo](https://cloud.githubusercontent.com/assets/13642402/9429498/35ffd74a-49d2-11e5-90d4-d794c181f42d.png)

**Comment by davecocoa** [comment]
Hi @JimVercoelen . Drawing a diagram was a great idea! However I don't think your arrows are quite accurate. If you want to join up all the streams conceptually, I would describe it like this:

`stdin`  ->  `tr`(input side) -> (some kind of magic) -> `loud`(output side) -> (your uppercase function) -> `loud`(input side) -> (some more magic) -> `tr`(output side) -> `stdout`

To clarify, for the two magic parts, trumpet is filtering the data it was given for elements with class `.loud`, and only passing those elements to `loud`, which outputs them into your uppercase function, once `loud` gets the elements back, it magically gives them back to `tr`, which stitches them back in place with the rest of the elements before everything gets streamed back to `stdout`.

I hope that helps!

**Comment by JimVercoelen** [comment]
Hi @davecocoa, thanks for your fast response even when te topic was a bit old.
Thanks for your diagram, it makes it much more clear!
Was reading your answers over and over and even though, it wasn't that clear to me (just started learning node etc).
So thanks a lot!

**Comment by smrutimandal** [comment]
Hey @davecocoa,

I am a complete newbie in this and not sure if just going through your first comment in this thread is cheating. ;)

I was struggling with this for the whole day and you made it seem so easy. Thanks a lot mate.

My solution if anyone is interesting:

var trumpet = require('trumpet');
var map = require('through2-map')
var tr = trumpet();
process.stdin.pipe(tr);

var elemstream = tr.select('.loud').createStream();
elemstream.pipe(map(function (input) {
	return input.toString().toUpperCase();
})).pipe(elemstream)

tr.pipe(process.stdout);

P.S. The documentation of trumpet where it says tr.select(selector) returns object element, I guess can add that the return is also a stream. Things can be slightly easier for newbies like me.

**Comment by davecocoa** [comment]
I'm glad it helped! I don't think reading comments is cheating, it's up to you to decide whether you have energy to keep trying or need some help with different examples/explanations.

**Comment by finferflu** [comment]
I've been struggling with this too, and posted my question in a [different discussion](https://github.com/nodeschool/discussions/issues/1397) (I hadn't yet seen this one). I've received help from a user in that discussion, but this discussion is also very illuminating. However, there is still something that irks me: if I simply create the `loud` stream by using the `trumpet.select()` and `trumpet.createStream()` methods, like so:

```javascript
var trumpet = require('trumpet')
var tr = trumpet()
var loud = tr.select('.loud').createStream()

process.stdin
    .pipe(tr)
    .pipe(process.stdout)
```

the resulting output from `trumpet` is:

```html
<html>
  <head>
    <title>beep boop</title>
  </head>
  <body>
    <p>
      Four score and several years ago, our fathers bought four swiss
      continents, a new vacation, covered in liberty.
      And predicated to the preposition that tall men created a
      sequel.
    </p>

    <p>
      How we are offstage in a great livermore, testing weather stations, or any
      station so conceived in altogether fitting and little note, nor long
      remember, they who fought here and take increased devoation,
      that <span class="loud">
```
(this is of course also true when I don't pipe the output from the `through2` function back to `loud` in this exercise).

Namely, the content is trimmed at `.loud`, and we also lose the closing `p`, `body`, and `html`. According to what's been discussed above, it would have made more sense to me if only the inner HTML of the `.loud` element was missing, but as it is this makes no sense to me. Does anybody have an explanation for this?

Thanks a lot for your time :)

**Comment by davecocoa** [comment]
I am not an expert on the internals of `trumpet`, so take this answer with a certain amount of uncertainty. I believe `trumpet` is attempting to stream output in order, therefore it cannot output any content *after* the `.loud` element until it has output the content *inside* the `.loud` element. However, it will *never* be able to output the content of the `.loud` element because you have not setup the streams needed to achieve that.

To be honest, I expected that program to throw an error waiting for the streams to close but I imagine they have some internal timeout after they've waited for data and consider it unlikely that any more will arrive.

**Comment by finferflu** [comment]
@davecocoa, thanks for your explanation, it makes a lot of sense :)

**Comment by iam-peekay** [comment]
@davecocoa super helpful explanation. thank you! :)

**Comment by paroxyzm** [comment]
@davecocoa super helpful explanation. thank you! :)
This thread should be mentioned in trumpet docs! :)

**Comment by crw** [comment]
I would strongly suggest working comment: https://github.com/nodeschool/discussions/issues/346#issuecomment-47089781 into the instructions for this problem in some way. That explanation was clear, concise, but very specific to trumpet. I do not know how someone would intuit the answer without that explanation or pouring through the trumpet docs.

**Comment by mcmonkeys1** [comment]
...does no one else think this is totally unnecessarily over-complicated? And the answer given is not even correct to the given question, i.e. selectAll not select. Would anyone of you use something as unintuitive in production code? This is all just bad coding.

**Comment by bengovernment** [comment]
@mcmonkeys1 ,  I'm starting to agree with you... the syntax surrounding trumpet seems really janky and unintuitive.

**Comment by pcuci** [comment]
Haven't looked into trumpet's internals, but I suspect the first *(some kind of magic)* is conceptually equivalent to https://en.wikipedia.org/wiki/Tee_(command)

**Comment by jphedley** [comment]
In hindsight, I agree & what for. At some point you'll use a Template. Me personally, it's http://cycle.js.org/, http://koajs.com/, http://reactivex.io/ (Generalized Streams) & too some extent, React & Angular 2.

**Comment by beeva-fernandobordallo** [comment]
¡¡¡Hey there!!!

First of all I'd like to say 'THANK GOD I'M NOT ALONE HERE!'. This exercise is well beyond logic. After a long struggle with it I reached a solution that I hope can help others understand 'the magic' behind trumpet.

(Node 4.4.4 used)

**FIRST STEP** - Check out the trumpet Stream

```
'use strict';

const trumpet = require('trumpet');
const fs = require('fs');

const tr = trumpet();

fs.createReadStream('test-inputs/html-input.html').pipe(tr).pipe(process.stdout);
```
This prints out the HTML input **as it comes**. No changes to it. From this we verify the duplex nature of the *tr* stream.

**SECOND STEP** - Create the second stream using the example given in the exercise example

```
const trumpet = require('trumpet');
const fs = require('fs');

const tr = trumpet();

fs.createReadStream('test-inputs/html-input.html').pipe(tr).pipe(process.stdout);

const selectStream = tr.select('div').createStream();

selectStream.on('data', (buffer) => {
	console.log(`Select Stream Data: ${buffer.toString()}`);
});
```
In this case I used 'div' as the selector. As I had created a new stream I expected:
a. The input html to be printed out the same as before, complete and without changes
b. The 'div' or 'div's found by the selector to be printed out separately on every `data` event received from the newly created stream.

THIS WAS NOT THE CASE!! The HTML streamed from `tr` to `stdout` STOPPED when the first element that matched was encountered. Hmmmmm ....

**THIRD STEP** - Try out one of the use cases given in the official [Trumpet Module Docs](https://www.npmjs.com/package/trumpet). More precisely, the first example found. It seemed a good one as it REPLACED the innerHTML of the matched element for what was written.

```
const trumpet = require('trumpet');
const fs = require('fs');

const tr = trumpet();

fs.createReadStream('test-inputs/html-input.html').pipe(tr).pipe(process.stdout);

const selectWriteStream = tr.select('div').createWriteStream();

selectWriteStream.end('HELL IS REAL');
```
Indeed, worked as expected. Not the solution, but it kind of shows a little of the inner workings of trumplet. What happens is:
1. HTML Stream piped to trumpet
2. Trumpet checks if the selector is matched
    - IF NO MATCH -> pipe data chunk to process.stdout
    - IF MATCH -> Creates a new writeable stream which we write to and close (all in one with `.end()`) and which is afterwards piped to process.stdout

As shown in the documentation example, this ONLY SELECTS THE FIRST MATCH. In an attempt to exchange all matches and get closer to the final solution I went on to mixing the first documentation example with the second.


**FOURTH STEP** - Repeat third step but using the selectAll option trumpet gives us

```
const trumpet = require('trumpet');
const fs = require('fs');

const tr = trumpet();

fs.createReadStream('test-inputs/html-input.html').pipe(tr).pipe(process.stdout);

const selectWriteStream = tr.selectAll('div', (div) => {
	div.createWriteStream().end('HELL IS REAL');
});
```
¡¡¡Closer!!! Ok. Up until here things seem acceptable. What happens is the same as we found happened in the THIRD STEP but we find ALL MATCHING ELEMENTS.


**FIFTH STEP** - Joining examples one and two from the documentation with the example given by the nodeschool exercise.

What we will do is, instead of creating a writeable stream (`div.createWriteStream()`) we will create a DUPLEX Stream with `.createStream()`:

```
const trumpet = require('trumpet');
const fs = require('fs');

const tr = trumpet();

fs.createReadStream('test-inputs/html-input.html').pipe(tr).pipe(process.stdout);

const selectWriteStream = tr.selectAll('span', (span) => {
	const elementStream = span.createStream();

	elementStream.on('data', (buffer) => {
		console.log(buffer.toString());
	});
	elementStream.end(); // VERY IMPORTANT TO CLOSE THIS STREAM
});
```
Here is where I understood what was happening. With the selectAll we retrieve a MATCH and we can create a stream (readable, writeable or duplex) out of it. In the previous case we created a writeable and overwrote the data (closing the stream as we did so). Just for kicks I created a readable stream (`span.createReadStream().on('data', (buf)=>{ console.log(buf.toString() )});`) and found that neither the readOnly or the writeOnly example blocked the printing to the stdout. 

HOWEVER, if you create a duplex stream as shown in this last example, if you don't close the stream the printing to stdout WILL STOP! So here is what is happening:

1. HTML is streamed into the `tr` stream
2. Trumpet checks if the selector is matched
    - IF NO MATCH -> pipe data chunk to process.stdout
    - IF MATCH -> Creates a new duplex stream which we can read the data from, can write data to and MUST CLOSE for it to be piped to process.stdout

This gets us to the grand finale!!!!

**SIXTH AND FINAL STEP** - Read data event from our matchDuplexStream, transform it and write it back to itself for the printing to continue

```
const trumpet = require('trumpet');

const tr = trumpet();

process.stdin.pipe(tr).pipe(process.stdout);

const selectWriteStream = tr.selectAll('.loud', (loud) => {
	const elementStream = loud.createStream();
	elementStream.on('data', (buf)=>{
		elementStream.end(buf.toString().toUpperCase());
	});
});
```
The final step seen here where we use .end inside the .on('data') event of the stream is in esence the same thing as stream.pipe(stream) with the addition of transforming the data buffer on the way.

**Comment by TigerLeep** [comment]
I found this thread probably through the same circumstances everyone else did.  I hit NodeSchool's STREAM ADVENTURE workshop, exercise HTML STREAM and was thoroughly confused by the trumpet package.  I *think* I have an understanding now.

- We connect the stdin stream's output to the trumpet stream's input.
- We connect the trumpet stream's output to the stdout stream's input.
- We ask the trumpet stream to create a NEW stream from elements with the class ".loud".
  - The loud stream that trumpet stream creates is a "special" stream:
    - It is created PRELOADED with the .loud elements
    - It is created with it's input internally connected back into the trumpet stream so anything received by the loud stream's input gets routed back into the trumpet stream.
- We connect the output of the loud stream to the input of the through2 stream (which converts to uppercase) before sending to its output.
- We connect the output of the through2 stream to the input of the loud stream (which is internally connected back into the trumpet stream.

The "AHA!" moment for me was when I realized the loud stream that the trumpet stream created for us DOES NOT PASS ITS INPUT THROUGH TO ITS OUTPUT!  It comes "preloaded" with data in its stream that it sends directly to its output.  It did not get this data from its input.  Data that it receives on its input it DOES NOT SEND TO ITS OUTPUT. Instead it sends that data back into the trumpet stream (which created it in the first place with its odd-ball wiring).

```
            -----------------------
            |   TRUMPET STREAM    |
STDIN--->---|->------      ----->-|--->STDOUT
            |        \    /       |
            -----------------------
                       \/     
                       /\
                ----------------    ----------------
                |    /    \    |    |              |
            |-->|---        ->-|--->|------------>-|-->|
            |   | LOUD STREAM  |    |   THROUGH2   |   |
            |   ----------------    ----------------   |
            |-<-------------------<------------------<-|
```

Inside TRUMPET STREAM there's also some magic for combine what it receives from LOUD STREAM's input back with the rest of TRUMPET STREAM's input that didn't get "materialized" inside LOUD STREAM when it was created.

Hopefully this helps to make it clearer for others.  Also, I discovered I'm pretty good at monospace font (aka ASCII) art.  LOL! :)

**Comment by Jeet007** [comment]
Hi,

I did the same problem but with different approach. Instead of piping it back to loud, I piped it directly to process.stdout via through since through is also a transform stream. 

Here is the code: 

var trumpet  = require('trumpet')
var through = require('through2')
var fs = require('fs')
var write = function(buffer,_,next) {
this.push(buffer.toString('utf-8').toUpperCase()+"\n")
next()
}

var through2 = through(write)

var tr = trumpet()

fs.createReadStream('./input.html').pipe(tr)

var html = tr.select('.loud').createReadStream().pipe(through2).pipe(process.stdout)

Ofcourse, stream-adevnture did not approve of it due to different syntax. But I would like to get feedback. Is this a correct way to do this?
